libass/ass_alloc.c

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (C) 2020 libass contributors
+ *
+ * This file is part of libass.
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "config.h"
+#include "ass_compat.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "ass_alloc.h"
+#include "ass_utils.h"
+
+// strdup and strndup are only part of ISO C since C2x; libass only requires C99
+#ifndef HAVE_STRDUP
+char *ass_strdup_replacement(const char *str)
+{
+    size_t len    = strlen(s) + 1;
+    char *new_str = ass_malloc(len);
+    if (new_str)
+        memcpy(new_str, str, len);
+    return new_str;
+}
+#define strdup ass_strdup_replacement
+#endif
+
+#ifndef HAVE_STRNDUP
+char *ass_strndup_replacement(const char *s, size_t n)
+{
+    char *end = memchr(s, 0, n);
+    size_t len = end ? end - s : n;
+    char *new = len < SIZE_MAX ? ass_malloc(len + 1) : NULL;
+    if (new) {
+        memcpy(new, s, len);
+        new[len] = 0;
+    }
+    return new;
+}
+#define strndup ass_strndup_replacement
+#endif
+
+
+// Following macros expect file, line and lib to be defined and initialised
+#ifndef NDEBUG
+    #define ASS_ALLOC_TEST(ptr) \
+        if (!ptr) \
+            ass_msg(lib, MSGL_ERR, "Alloc fail in %s at line %u!\n", file, line)
+#else
+    #define ASS_ALLOC_TEST(ptr) do { } while (0)
+#endif
+
+#define ASS_RELAY_ALLOC(fun, ...) \
+    void *new_ptr = (fun)(__VA_ARGS__);  \
+    ASS_ALLOC_TEST(new_ptr); \
+    return new_ptr
+
+void *ass_malloc_fun(const char *file, const unsigned line, ASS_Library *lib,
```

Comment:
I think there's no much sense in duplication of all libc allocation functions as is. It's enough to implement a single `ass_realloc()` plus bunch of inline functions or macros in the header.

Reply:
`realloc` alone may be able to emulate all other standard allocation functions, but as before libc's implementations are potentially better optimised and it is simpler to just relay to standard functions than to reimplement them.
If for a future allocator API, you'd prefer the user to only needing to supply one realloc and free implementation, I'd prefer to leave the emulation of the other allocating functions to such a future PR.